src/vueWrapper.ts

```diff
@@ -257,6 +258,26 @@ export class VueWrapper<T extends ComponentPublicInstance>
 
     this.__app.unmount()
   }
+
+  async trigger(
+    eventString: Parameters<T['$emit']>[0],
```

Comment:
I don't think this works in all cases. This is going to accept only the parameters of the first `$emit` function, whereas there might be several signatures possible no?

`emits:` accepts an array with several events, or an object with several keys. It would be great to have tests showing that all these cases work (and I'm suspecting this is not the case).

Reply:
Yes, this tripped me up as well, this would ideally be `Parameters<T['$emit']>[number]` but then I don't get the typehint.

> It would be great to have tests showing that all these cases work

I'm not aware how you could test the arguments of the trigger method in type testing.

Reply:
I would add a few calls to trigger in `wrapper.d-test.ts`, like:
```ts
const CompWithEmitsAsObject = defineComponent({ emits: { event1: () => true, event2: () => true }});
const wrapper = mount(CompWithEmitsAsObject);
wraper.emit('event1');
wraper.emit('event2');
// @ts-expect-error 'other' is not defined as an emitted event
wrapper.emit('other');
```
Same thing for events defined in an array

Reply:
Ah, I got it now.
vueWrapper now also accepts dom event name and string so this is not applicable anymore

Reply:
@pikax Maybe you can help here? We'd like to type `eventString` to accept the possible emitted events by the component. As I explained above, I'm afraid `Parameters<T['$emit']>[0]`, but you probably know best.

Reply:
@cexbrayat  I would extract the `Emits` like this:

```ts
const x = defineComponent({
  emits: {
    click: (n: number) => {},
    input: (b: string) => b.length > 1
  }
})
declare const t: typeof x extends DefineComponent<
  any,
  any,
  any,
  any,
  any,
  any,
  any,
  infer E
>
  ? E
  : undefined

t.click(1)
t.input('1')

//@ts-expect-error
t.click('2')
//@ts-expect-error
t.input(1)
```

Reply:
I don't see the issue with `Parameters<T['$emit']>[0]` it's always going to be the first parameter of the emit function and it is correctly inferring both:
```ts
emits: {
    click: (n: number) => {},
    input: (b: string) => b.length > 1
  }
```
and
`emits: ['some', thing']`

when writing `wrapper.trigger('the event')`

Still cannot be tested as it accepts `string` too here.

Reply:
Oh, I missed the edit of your comment.

Yeah I suspected it was not enough. I guess you can take a look into how the `emit` function of Vue itself is typed, but this is probably quite complex to handle properly.

Or maybe @pikax can give us a hand?

If this is too complex we can simply remove it and just use string. In that case, I guess all the overrides in the base class will be enough, and the code in both wrappers can be removed.

Reply:
> I don't see the issue with `Parameters<T['$emit']>[0]` it's always going to be the first parameter of the emit function and it is correctly inferring both:

For events we need both event name and argument typed.

> I was wrong it was inferring the object syntax because it was click and input which is also a dom event name.

If the user declared `click` and `input` that component will not trigger the [DOM event](https://v3.vuejs.org/api/options-data.html#emits). - "Events listed in the emits option will not be inherited by the root element of the component and also will be excluded from the $attrs property." 






> For reference `E extends EmitsOptions` which is:
> `export declare type EmitsOptions = ObjectEmitsOptions | string[];` and `ObjectEmitsOptions` is a object with functions as you might have guessed.
> Have you got some guidance for correcting this type extraction?


The `emit` function type used in the setup context is [EmitFn<E>](https://github.com/vuejs/vue-next/blob/24e9efcc21dd517fddf8251873a06d71830e46eb/packages/runtime-core/src/componentEmits.ts#L24-L35) where's `E` is the `emit` object